exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +319,264 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+	for k, v := range mp.Spec.Template.Labels {
+		machine.Labels[k] = v
+	}
+
+	// Enforce that the MachinePoolNameLabel and ClusterNameLabel are present on the Machine.
+	machine.Labels[clusterv1.MachinePoolNameLabel] = capilabels.MustFormatValue(mp.Name)
+	machine.Labels[clusterv1.ClusterNameLabel] = mp.Spec.ClusterName
+
+	return machine
+}
+
+// infraMachineToMachinePoolMapper is a mapper function that maps an InfraMachine to the MachinePool that owns it.
+// This is used to trigger an update of the MachinePool when a InfraMachine is changed.
+func (r *MachinePoolReconciler) infraMachineToMachinePoolMapper(ctx context.Context, o client.Object) []ctrl.Request {
+	log := ctrl.LoggerFrom(ctx)
+
+	machinePoolList := &expv1.MachinePoolList{}
+	labels := o.GetLabels()
+	selector := map[string]string{}
+	clusterName, ok := labels[clusterv1.ClusterNameLabel]
+	if !ok {
+		selector = map[string]string{clusterv1.ClusterNameLabel: clusterName}
+	}
+
+	// TODO: Document that the label selector must be the hash of the MachinePool name once we confirm this works.
+	if poolNameHash, ok := labels[clusterv1.MachinePoolNameLabel]; ok {
+		if err := r.Client.List(ctx, machinePoolList, client.InNamespace(o.GetNamespace()), client.MatchingLabels(selector)); err != nil {
+			log.Error(err, "failed to get MachinePool for InfraMachine")
+			return nil
+		}
+
+		for _, mp := range machinePoolList.Items {
+			// TODO: Remove logs once we confirm this works
+			log.V(4).Info("Found MachinePool", "machinePool", mp.Name, "hash", capilabels.MustFormatValue(mp.Name))
+			if capilabels.MustFormatValue(mp.Name) == poolNameHash {
+				return []ctrl.Request{
+					{
+						NamespacedName: client.ObjectKey{
+							Namespace: mp.Namespace,
+							Name:      mp.Name,
+						},
+					},
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *MachinePoolReconciler) waitForMachineCreation(ctx context.Context, machineList []clusterv1.Machine) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// waitForCacheUpdateTimeout is the amount of time allowed to wait for desired state.
+	const waitForCacheUpdateTimeout = 10 * time.Second
+
+	// waitForCacheUpdateInterval is the amount of time between polling for the desired state.
+	// The polling is against a local memory cache.
+	const waitForCacheUpdateInterval = 100 * time.Millisecond
+
+	for i := 0; i < len(machineList); i++ {
+		machine := machineList[i]
+		pollErr := wait.PollUntilContextTimeout(ctx, waitForCacheUpdateInterval, waitForCacheUpdateTimeout, true, func(ctx context.Context) (bool, error) {
+			key := client.ObjectKey{Namespace: machine.Namespace, Name: machine.Name}
+			if err := r.Client.Get(ctx, key, &clusterv1.Machine{}); err != nil {
+				if apierrors.IsNotFound(err) {
+					return false, nil
+				}
+				return false, err
+			}
+
+			return true, nil
+		})
+
+		if pollErr != nil {
+			log.Error(pollErr, "Failed waiting for machine object to be created", "Machine", klog.KObj(&machine))
```

Comment:
```suggestion
```
Let's drop this log. I think returning the error is enough (the error will be logged by CR eventually)